### ISSUE TITLE
cronjobs_log table getting flooded with entries

### ISSUE THREAD

**Issue by travisbotello** [Jako/CronManager#62]
I have the following setup:

![One cron job that runs every 24 hours.](https://github.com/user-attachments/assets/34a0df4e-6e93-4048-bdef-8165f196050a)

I am triggering the cronjob every minute via crontab:

php /var/www/public_html/assets/components/cronmanager/cron.php

When I check the cronjobs_log table it's getting constantly flooded and my server is under heavy load. It's generating around 10k entries per 5 minutes. For whatever reason the cronjob is triggered many times every minute although I set it to 1440 minutes:

![Screenshot 2024-08-14 at 17 26 30](https://github.com/user-attachments/assets/5e049770-790d-48cf-bc94-9ad54f507259)

I am running MODX 2.8.7 with CronManager 1.3.2

**Comment by travisbotello** [comment]
I just downgraded to 1.3.1 and the issue is gone...

**Comment by Jako** [comment]
What is written in the table?

**Comment by travisbotello** [comment]
> What is written in the table? Can you mail me the content of your cronjobs table?

It‘s logging the default output of my snippet. So it appears to be constantly triggering it. Same timestamp for a bunch of entries, chunked. Will send via email once I am back at the desk.

**Comment by Maef** [comment]
We had the same issue last week. Downgrading fixed the endless loop. The while loop of commit https://github.com/Jako/CronManager/commit/2982e6b2a67b1773d5ea93780734b71002cca38a must be causing this. Two snippets run at an 60 and 240 minutes interval.

**Comment by Maef** [comment]
@Jako If I replace line 121 in `cron.php`  with `$cronjob->set('lastrun', date('Y-m-d H:i:s'));` it works. Your line 56 is also necessary. Setting the lastrun property with the variable $rundatetime seems to be part of the endless loop issue.

Can you test/confirm?

**Comment by Jako** [comment]
Strange, 'lastrun' is not not in the where of the cronjobs query. So why does this change anything?
 
 Can you try it with setting the cacheFlag to 0 in line 124:
 
 ```$cronjob->save(0);```

**Comment by Maef** [comment]
Still endless loop with `while ($cronjob = $modx->getObject('modCronjob', $c, false)) {` and `$cronjob->save(0);`

**Comment by Jako** [comment]
I really don't know why this happens. I suppose that a database caching layer changes the query result if the cronjob record is saved with a new value in lastrun every loop. 

I will rewrite the code without that while loop. It has to use a `started` column to catch currently running jobs and some code  that clears this `started` column after a while to avoid locked jobs where the `started` column is not cleared after the job is finished.

**Comment by bYemma** [comment]
I have the same issue

**Comment by travisbotello** [comment]
> I really don't know why this happens. I suppose that a database caching layer changes the query result if the cronjob record is saved with a new value in lastrun every loop.
> 
> I will rewrite the code without that while loop. It has to use a `started` column to catch currently running jobs and some code that clears this `started` column after a while to avoid locked jobs where the `started` column is not cleared after the job is finished.

I suspect the change to the `$rundatetime` definition to be the source of error inside the while loop:

```php
$rundatetime = ($cronjob->get('lastrun')) ? $cronjob->get('lastrun') : date('Y-m-d H:i:s');
```

With previous version it was defined outside of the loop as `$rundatetime = date('Y-m-d H:i:s');`. So it was always referencing the current date. With the current implementation it can hold an old date of the `lastrun` column, if set.

This could lead to an issue when we are saving the `nextrun` on this cronjob. The `nextrun` is crucial for the condition to fetch outstanding cronjob and to eventually break the loop:

```php
    $cronjob->set('lastrun', $rundatetime);
    $cronjob->set('nextrun', date('Y-m-d H:i:s', (strtotime($rundatetime) + ($cronjob->get('minutes') * 60))));
```

Isolated test case:

```php
<?php

$nextrun = '2024-08-24 10:00:00';
$lastrun = '2024-07-24 10:00:00';
$minutes = 180;

$rundatetime = $lastrun ? $lastrun : date('Y-m-d H:i:s');

$lastrun = $rundatetime;
$nextrun = date('Y-m-d H:i:s', (strtotime($rundatetime) + $minutes * 60));

echo $nextrun;
```

This returns `2024-07-24 13:00:00`. If you update the values it returns `2024-07-24 13:00:00` and so back and so forth...so the loop continues running and running although it should've been stopped after the first run.

**Comment by Jako** [comment]
Phew, that's it!

**Comment by Jako** [comment]
Please try this fix: https://github.com/Jako/CronManager/commit/ab2064e505ec2169e1767b67b82836f5c9fde4db

**Comment by Jako** [comment]
This is fixed with 1.4.0

**Comment by Jako** [comment]
Version 1.4.0 is available in the MODX extras repository.
